Metadata for input/output formats in model file proto.

pritamdamania87 opened the pull request:
Summary:
We would like to extend the model file format to include strongly type, semantic information
about the model inputs and outputs.

The goal is for a user to be able to consider a model file like a function with
a well defined API describing what the inputs and outputs would be.

Differential Revision: D13009915

houseroad commented on caffe2/proto/torch.proto:
```diff
@@ -84,5 +84,7 @@ message ModelDef {
-  // future: add a way to provide additional meta-data
+  // Serialized metadata for input and outputs (e.g: expected types, shapes and
+  // semantic information).
+  optional RecordRef module_metadata = 6;
```
Nit: rename to predictor_metadata or metadata?

pritamdamania committed:
Metadata for input/output formats in model file proto. (#15252)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15252

We would like to extend the model file format to include strongly type, semantic information
about the model inputs and outputs.

The goal is for a user to be able to consider a model file like a function with
a well defined API describing what the inputs and outputs would be.

Reviewed By: dzhulgakov

Differential Revision: D13009915

fbshipit-source-id: efc813b30a294e6059cb588c605347612b6b2ed8
